pyqtorch/hamiltonians/evolution.py

```diff
@@ -394,24 +438,82 @@ def spectral_gap(self) -> Tensor:
         spectral_gap = torch.unique(torch.abs(torch.tril(diffs)))
         return spectral_gap[spectral_gap.nonzero()]
 
+    def _time_evolution(
+        self, values: dict[str, Tensor] | ParameterDict | None = None
+    ) -> Tensor:
+        """
+        Get the evolution from parameter values.
+        Arguments:
+            values: Values of parameters.
+        Returns:
+            The time evolution.
+        """
+        values = values or dict()
+        if isinstance(self.time, str):
+            pname = self.time
+            # note: GPSR trick when the same param_name is used in many operations
+            if self._param_uuid in values.keys():
+                pname = self._param_uuid
+            time_evolution = values[pname]
+        elif isinstance(self.time, ConcretizedCallable):
+            time_evolution = self.time(values)
+        else:
+            time_evolution = self.time
+        return time_evolution
+
     def _forward(
         self,
         state: Tensor,
         values: dict[str, Tensor] | ParameterDict | None = None,
         embedding: Embedding | None = None,
     ) -> State:
         values = values or dict()
```

Comment:
Yeah I would suggest to default it as early as possible so that we don't repeat that check further down the abstraction.